### PR TITLE
Support CKM_AES_CTR

### LINKED ISSUE #69: CKM_AES_CTR support
Only the mechanism type is supported.

### PR DESCRIPTION
**Pull request by emillon**
This adds a high level type with bits and block (invariants are not
checked), and a low level corresponding to the C representation.

See <http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850499>

Closes #69

### CODE REVIEW

**Comment by emillon** [comment]
As an extra check, I ran the following program on softhsm2 to compare CTR output of the null plaintext with ECB encryption, and it's indeed equal. I don't think it's worth adding to the code base because it relies on a GCM soft token, and it tests more the token than the bindings anyway.

```ocaml
let make_ecb_res (module S:P11_driver.S) session key nonce =
  let mechanism = P11.Mechanism.CKM_AES_ECB in
  let encrypt = S.encrypt session mechanism key in
  let plain1 = Printf.sprintf "%s\x00\x00\x00\x00\x00\x00\x00\x01" nonce in
  let plain2 = Printf.sprintf "%s\x00\x00\x00\x00\x00\x00\x00\x02" nonce in
  let cipher1 = encrypt plain1 in
  let cipher2 = encrypt plain2 in
  cipher1 ^ cipher2

let make_ctr_res (module S:P11_driver.S) session key nonce =
  let bits = 64 in
  let block = Printf.sprintf "%s\x00\x00\x00\x00\x00\x00\x00\x01" nonce in
  let param = P11.AES_CTR_params.make ~bits ~block in
  let mechanism = P11.Mechanism.CKM_AES_CTR param in
  let plain = String.make 32 '\x00' in
  S.encrypt session mechanism key plain

let run ~dll ~slot_id ~pin =
  Pkcs11_log.set_logging_function prerr_endline;
  let (module S) =
    P11_driver.load_driver
      ?log_calls:None
      ?on_unknown:None
      ~dll
      ~use_get_function_list:`Auto
  in
  S.initialize ();
  let slot = match S.get_slot slot_id with
    | Ok s -> s
    | Error e -> failwith e
  in
  let session = S.open_session ~slot ~flags:P11.Flags._CKF_SERIAL_SESSION in
  S.login session P11.User_type.CKU_USER pin;
  let template =
    [ P11.Attribute.Pack (P11.Attribute_type.CKA_ENCRYPT, true)
    ; P11.Attribute.Pack (P11.Attribute_type.CKA_CLASS, P11.Object_class.CKO_SECRET_KEY)
    ; P11.Attribute.Pack (P11.Attribute_type.CKA_KEY_TYPE, P11.Key_type.CKK_AES)
    ; P11.Attribute.Pack (P11.Attribute_type.CKA_VALUE_LEN, Unsigned.ULong.of_int 16)
    ]
  in
  let mechanism = P11.Mechanism.CKM_AES_KEY_GEN in
  let key_handle = S.generate_key session mechanism template in
  let nonce = "AAAABBBB" in
  let ecb_res = make_ecb_res (module S) session key_handle nonce in
  let ctr_res = make_ctr_res (module S) session key_handle nonce in
  Printf.printf
    "ECB: %S\nCTR: %S\neq:  %B\n"
    ecb_res
    ctr_res 
    (ecb_res = ctr_res)

let () =
  match Sys.argv with
  | [| _ ; dll ; slot_string ; pin |] ->
    let slot_id = P11.Slot.Index (int_of_string slot_string) in
    run ~dll ~slot_id ~pin
  | _ ->
      invalid_arg "Usage: ctrtest <dll> <slot> <pin>"
```

File: test/test_p11_aes_ctr_params.ml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      got
+  in
+  "of_yojson" >:::
+  [ "normal" >:: test
+      (`Assoc
+         [ ("bits", `String "64")
+         ; ("block", `String "AAAABBBBCCCCDDDD")
+         ]
+      )
+      (Ok (
```
Comment (bbc2) [inline test/test_p11_aes_ctr_params.ml]: The style is inconsistent here.

**Review by bbc2 — CHANGES_REQUESTED** [review]
👍 just some minor style issue.

**Comment by emillon** [comment]
I fixed the style, let me know what you think.

**Review by bbc2 — APPROVED** [review]

**Pull request merged**